### PR TITLE
Watchdog: Allow sending notifications via webhooks

### PR DESCRIPTION
**Pull request by felixoi**
Inspired by #3672 and #4198 this adds a way to send watchdog notifications to any JSON-compilant webhook service.

### CODE REVIEW

**Comment by felixoi** [comment]
In addition to this PR, I'm thinking about adding support for the health status webhooks of every service monitored by watchdog (to use services like healthchecks.io). This would result in a lot of new config variables. Not sure how to handle this.

File: update.sh [hunk 1]
```diff
@@ -546,6 +548,20 @@ for option in ${CONFIG_ARRAY[@]}; do
...
+      echo '# Send notifications to a webhook URL that receives a POST request with the content type "application/json".' >> mailcow.conf
+      echo '# You can use this to send notifications to services like Discord, Slack and others.' >> mailcow.conf
+      echo '#WATCHDOG_NOTIFY_WEBHOOK=https://discord.com/api/webhooks/XXXXXXXXXXXXXXXXXXX/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' >> mailcow.conf
+    fi
+  elif [[ ${option} == "WATCHDOG_NOTIFY_WEBHOOK_BODY" ]]; then
+    if ! grep -q ${option} mailcow.conf; then
+      echo "Adding new option \"${option}\" to mailcow.conf"
+      echo '# JSON body included in the webhook POST request. Needs to be in single quotes.' >> mailcow.conf
+      echo '# Following variables are available: SUBJECT, BODY' >> mailcow.conf
+      echo '#WATCHDOG_NOTIFY_WEBHOOK_BODY=\'{"username": "Mailcow Watchdog", "content": "**${SUBJECT}**\n${BODY}"}\'' >> mailcow.conf
```
Comment (MAGICCC) [inline update.sh]: It's mailcow not Mailcow

File: generate_config.sh [hunk 2]
```diff
@@ -338,6 +338,13 @@ USE_WATCHDOG=y
 #WATCHDOG_NOTIFY_EMAIL=a@example.com,b@example.com,c@example.com
 #WATCHDOG_NOTIFY_EMAIL=
 
+# Send notifications to a webhook URL that receives a POST request with the content type "application/json".
+# You can use this to send notifications to services like Discord, Slack and others.
+#WATCHDOG_NOTIFY_WEBHOOK=https://discord.com/api/webhooks/XXXXXXXXXXXXXXXXXXX/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+# JSON body included in the webhook POST request. Needs to be in single quotes.
+# Following variables are available: SUBJECT, BODY
+#WATCHDOG_NOTIFY_WEBHOOK_BODY='{"username": "Mailcow Watchdog", "content": "**${SUBJECT}**\n${BODY}"}'
```
Comment (MAGICCC) [inline generate_config.sh]: It's mailcow not Mailcow

**Review by MAGICCC — CHANGES_REQUESTED** [review]

**Comment by felixoi** [comment]
@MAGICCC fixed.

**Comment by osos** [comment]
may I suggest that somehow a differentiation in severity of the "error" is implemented and however for the hooks.

Maybe

- notify_error(subject, body)
- notify_warning(subject, body)
  - not sure which events are warnings
- notify_info(subject,body)
  - e.g. ip-ban

**Comment by felixoi** [comment]
I've a lot things in mind that can be improved here like healthchecks (send a health ping ever x seconds a service is healthy and a fail ping to instantly fail), multiple webhooks for different services, ... I'd like to get the basic webhook in first and then improve this whole monitoring notification topic afterwards in multiple PRs.

**Comment by MAGICCC** [comment]
Just tested a bit with the Pushover API and works fine:
image




Does it makes sense to add a `WATCHDOG_WEBHOOK_VERBOSE` variable which adds a -v in the curl command?

You need to add in docker-compose the env vars in the watchdog image too:
```
 - WATCHDOG_NOTIFY_WEBHOOK=${WATCHDOG_NOTIFY_WEBHOOK}
 - WATCHDOG_NOTIFY_WEBHOOK_BODY=${WATCHDOG_NOTIFY_WEBHOOK_BODY}
```

**Comment by felixoi** [comment]
Verbose option is pending because of #5081

**Comment by MAGICCC** [comment]
Hi @felixoi are you still working on this?

**Comment by felixoi** [comment]
@MAGICCC Yeah need to fix this up. When I started this I tried out mailcow and could just test this things on my Hetzner VM but now I'm using it in production so I can't test it there anymore. I may spin up a second instance to finish this..

**Comment by MAGICCC** [comment]
Nice to hear this!

**Comment by osos** [comment]
how is this going?

a couple of times we had the postfix to crash. The watchdog could send me an email.

**Comment by felixoi** [comment]
I'm really sorry that I haven't been able to finish this yet. My time is very limited at the moment and I would need to set up a new development environment. If anyone wants to finish this (I think just the verbose option is missing and tests if it works correctly if added), please feel free to do so.

**Comment by DerLinkman** [comment]
No promise but we might finish it. But also for that: no eta just looking into things which are nice to have.

**Comment by MAGICCC** [comment]
Maybe I have time at the weekend to finish this. As seen above it's already working, except a verbose option

**Comment by MAGICCC** [comment]
Thanks @FreddleSpl0it for it. Didn't have time myself

**Pull request merged**